qaoa_OOP.py

```diff
@@ -0,0 +1,279 @@
+        
+        self.params[f'xL_d{self.depth}'] = self.g_params[ind].copy()
+        self.E[f'{self.depth}']          = self.g_values[ind]
+        self.best[f'{self.depth}']       = self.g_best_values[ind]
+
+    def get_energy_landscape(self):
+
+        print("Calculating energy landscape ...")
+
+        ranges = []
+        for i in range(self.q):
+
+            step = (self.params_ul[i] - self.params_ll[i])/self.params_n[i]
+            
+            ranges.append(slice(self.params_ll[i],self.params_ul[i],step))
+        
+        x0, fval, grid, Jout = optimize.brute(self.getval, ranges, full_output = True)
+
+        print("Calculating energy landscape done.")
+        # Flatten the parameters according to fortran ordering
+        self.params[f'x0_d{self.depth}'] = x0.flatten()
+        
+        # Transpose energy landscape to be in accordance with the ordering used in the
+        # runQAOA function used previously.
+        return Jout.T, x0.flatten()
+
+    def simulate_init(self, **simulation_args):
+        
+        # Unpack the simulation arguments:
+        
+        self.backend     = simulation_args['backend']
+        self.optmethod   = simulation_args['optmethod']
+        self.max_depth   = simulation_args['max_depth']
+        self.repeats     = simulation_args.get('repeats',1)
+        self.shots       = simulation_args.get('shots',  1)
+        self.noise_model = simulation_args.get('noise_model',None)
+
+        # Give lower and upper limits (ll, ul) as arrays rather than
+        # separately giving max and min for each variable, so that it is
+        # easier to change the number of variables later on
+        # Aslo, provide the number of steps for each variable as a similar array
+        
+        self.params_ll = simulation_args['params_ll']
+        self.params_ul = simulation_args['params_ul']
+        self.params_n  = simulation_args['params_n']
+
+        self.q         = self.params_ll.size
+
+        assert( self.params_ll.size == self.params_ul.size == self.params_n.size)
+        
+    def simulate(self, **simulation_args):
+
+        """
+        Simulation function for doing the QAOA algorithm.
+
+        Parameters
+        ----------
+        simulation_args : dict
+            Keyword arguments, must contain:
+                backend
+                optmethod
+                max_depth
+                params_ll : lower limits for each parameter to do brute search over
+                params_ul : upper limits -- " --
+                params_n  : number of points -- " --
+
+                
+        """
+
+        self.simulate_init(**simulation_args)
+        self.depth = 1
+        
+        # Calculate energy_landscape - global optimisation 
+
+        Elandscape, x0 = self.get_energy_landscape()
+
+        # Local optimisation
+
+        for rep in range(self.repeats):
+            print(f"Depth = {self.depth}, Rep = {rep + 1}")
+
+            res = optimize.minimize(self.getval,
```

Comment:
it is confusing that "res" is returned but never used

Reply:
Agree, made it more apparent what it does.